distribution/lib/Standard/Database/0.0.0-dev/src/Internal/IR/SQL_Expression.enso

```diff
@@ -64,3 +64,57 @@ type SQL_Expression
          parametrized by additional settings can use this field to pass that
          information to the code generator.
     Operation (kind : Text) (expressions : Vector SQL_Expression) (metadata : Any | Nothing = Nothing)
+
+    ## PRIVATE
+       The internal representation of a `WITH ... AS` CTE expression. A `Let`
+       value contains the "binding" of the CTE expression to a lexically scoped
+       table name, as well as the body expression that refers to the binding.
+
+       A `Let` value's body always contains at least one `Let_Ref` value with
+       the same `binder` identifier, unless the callback passed to the `let`
+       method ignored its argument.
+
+       Arguments:
+       - name: the optional user-provided name describing the expression bound
+         by the WITH ... AS clause
+       - binder: the lexically-scoped name of the CTE expression.
+       - bindee: the CTE expression.
+       - body: the larger expression that contains references to the CTE.
+    Let (name : Text) (binder : Text) (bindee : SQL_Expression) (body : SQL_Expression)
+
+    ## PRIVATE
+       The internal representation of a "reference" to a CTE expression.
+
+       A `Let_Ref` value always occurs within the `body` expression of an
+       enclosing `Let` value.
+
+       Database backends often evaluate sub-expressions of `DB_Table` and
+       `DB_Column` values, for example when querying the datanbase for an
+       expression type. `Let_Ref` values are sometimes evaluated in the
+       database without their enclosing `Let` values, in which case the
+       `binder` field refers to an unbound identifier. In these cases, the
+       original CTE expression in `standalone_expression` is used instead.
```

Comment:
Tbh I'm not completely happy about it, but not sure if we can do anything about it either.

Surely we cannot restructure to ensure that the queries are always ran in a proper context? I'm actually surprised that we sometimes evaluate just the subexpression.

I guess the workaround of keeping the full expression to fill-in as a fallback is okay, but it makes me feel uneasy in that we are sometimes executing a different query - what if the behaviour diverges? On the other hand, it probably shouldn't ever diverge in a correct SQL implementation so we are probably fine...

Reply:
Yeah, I figure that since we're already executing a sub-query out of its context, we already have this problem.

The bound value in `Let` and the standalone expression in `Let_Ref` are both generated in just one place, together, so it's unlikely that they will diverge.